### ISSUE TITLE
Document query modifiers

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#65929]
We should document what the [query modifiers](https://github.com/rust-lang/rust/blob/eb5ef813f0d6e3fe8edd3abb046a18f5b1a8cc48/src/librustc_macros/src/query.rs#L33-L62) mean in a more involved form than currently, ideally in rustc-guide or so, with a comment in src/librustc/query/mod.rs pointing to there.

**Comment by Mark-Simulacrum** [comment]
I've nominated this for compiler team so that we can discuss and hopefully assign someone to this, since I think it is likely an "easy" task for someone with the knowledge (perhaps @eddyb or @michaelwoerister) to give a quick summary which can then be put into docs by someone from @rust-lang/wg-learning  perhaps. 

Some specific questions:
 * What does forcing a query mean? Why is it desirable to not allow that?
 * Does eval_always mean the query is always run regardless of incremental? Does that mean we don't serialize it to disk? Why would we want this?
 * What are anonymous queries for? What gain is there to not having a dep node for the query itself?

**Comment by eddyb** [comment]
I think the names and defaults are suboptimal, so we might also want to clean that up first.
`no_force` is required on any query that is not keyed on `DefId` or `CrateNum`, so maybe it should be type-based instead?
`eval_always` should be seldom used and merits a more descriptive name (it seems to be used for some "input" queries AFAICT?)

**Comment by Mark-Simulacrum** [comment]
Perhaps, although I think figuring out what they mean is also worthwhile -- does no_force then have something to do with not being able to "pull out" the query key from the hash we store in the serialized data? (So, if I have a DefId query, should I _never_ mark it as no_force?)

**Comment by eddyb** [comment]
I just checked and no queries with a key type implementing `RecoverKey` have `no_force`.
Those types are `DefId`, `CrateNum` and `DefIndex`:
https://github.com/rust-lang/rust/blob/eb5ef813f0d6e3fe8edd3abb046a18f5b1a8cc48/src/librustc/dep_graph/dep_node.rs#L434-L454

So presumably this method could be on the more general `Key` trait and have a default implementation that always returns `None`, or maybe something slightly fancier to avoid doing work? (specialization is also an option, I guess, if we can't avoid using it)

Then we wouldn't need `no_force`, I would think? cc @Zoxc

**Comment by Zoxc** [comment]
@eddyb Is `always_runs` a better name for `eval_always`? =P

Input queries isn't really a thing. All inputs must be available when the query system is started. We have queries which use an input and only return part of it though, but that is a more general pattern.

**Comment by michaelwoerister** [comment]
> What does forcing a query mean? Why is it desirable to not allow that?

"Forcing" basically means executing the query (i.e. we are forcing the result to be in the cache). In the context of incremental compilation, we force a query if we cannot guarantee that it has the same result during the current compilation session. In particular:

- During incremental compilation, before executing a query provider, we first try to mark the corresponding `DepNode` as green, because if it is green, we can use the result from the on-disk-cache. Or we might not need the result at all because we are just checking that there was no change. 
- However, if we encounter a `DepNode` that cannot be marked as green (i.e. one of the query inputs has changed) then we need to "force" the corresponding query in order to find out if the result has changed or not. 
- In such a case we take the `DepNode` and try to reconstruct the corresponding query key from it. Unfortunately, in the current implementation, this is only possible if the query key is of type `DefId`. If the query key has a different type, we cannot reconstruct the query key and thus also cannot force the query. All such queries *have* to be marked as `no_force` (otherwise you'll probably get compilation errors). 
- So `no_force` really means "this query cannot be re-run from just the `DepNode` it corresponds to because we don't have enough information to reconstruct the query key".

**Comment by Mark-Simulacrum** [comment]
> So, if I have a DefId query, should I never mark it as no_force?

So the answer here is yes, right? From what I understand based on what you've said, a DefId query can always be rerun.

And no_force is as such not really an additional marker, as @eddyb mentions, but more so just a way to bypass some "smart" logic based on the query inputs being DefId or not. i.e., you should only add it if you have to, never for actual "reasons" so to speak.

**Comment by michaelwoerister** [comment]
@Mark-Simulacrum Yes, that sounds correct.

**Comment by eddyb** [comment]
> So `no_force` really means "this query cannot be re-run from just the `DepNode` it corresponds to because we don't have enough information to reconstruct the query key".

This makes me even more confident in the belief that `no_force` should be type-based and automatic.

> @eddyb Is `always_runs` a better name for `eval_always`? =P

That still doesn't explain *why* you'd ever want that on a query, from my quick skimming it looks like it only exists on queries that read pre-filled data from `TyCtxt` (hence me calling them "input queries").

> Input queries isn't really a thing. All inputs must be available when the query system is started. We have queries which use an input and only return part of it though, but that is a more general pattern.

What about the queries that expose the inputs available when the query system is started? My guess was that `eval_always` was specifically used for them.

I would prefer a name that only made sense for whatever `eval_always` happens to be used for today, not "always run this query", which would seem to apply to anything if not further qualified.

**Comment by Zoxc** [comment]
> What about the queries that expose the inputs available when the query system is started?

These are only `eval_always` because that's slightly more efficient somewhere.

`eval_always` is used for 2 main reasons:
1) There some side effects that needs to be executed that incremental doesn't track (like writing to files or marking attributes a used). This is needed for correctness.
2) A query depends on pretty much the entire crate in which case we can skip recording its dependencies and just assume it changes. This is good for performance.

**Comment by eddyb** [comment]
What if we used two different names, to make it clearer?
1. `has_side_effects`
2. `depends_on_everything`

Bonus, you'd only rerun `has_side_effects` always, but `depends_on_everything` only if anything changed at all in the HIR.

**Comment by eddyb** [comment]
That sounds like:
* it should be rare
* it should only take effect in parallel mode
* it should be addressed ideally by parallelism in the query system itself
* last but not least, the name should indicate all that

**Comment by pnkfelix** [comment]
triage: leaving nominated and unprioritized; I want the team to discuss whether the discussion has reached the point where someone from @rust-lang/wg-learning  could pick it up and make a PR to the rustc-guide

**Comment by eddyb** [comment]
@pnkfelix I'd prefer changing both modifiers first, to an extent that would make documentation lower priority (since the remaining modifiers would never be meant to be used by most queries, only ones special one way or another).

We should still document the macro as a whole though (and a doc comment above the invocation in `src/librustc/query/mod.rs` wouldn't hurt).

**Comment by pnkfelix** [comment]
we [discussed](https://zulip-archive.rust-lang.org/131828tcompiler/07347weeklymeeting2019103154818.html#179537701) this in last week's meeting. I'm good with @eddyb's point that we should work on cleaning up the API itself before we invest effort in documenting it as it stands.

triage: P-medium. Removing nomination label.

**Comment by pnkfelix** [comment]
I am going to assign @eddyb for now, with the intent that they either do the API change they envisage, or find someone else to do it.
